### ISSUE TITLE
Jax decompiler

### ISSUE THREAD

**Issue by PierrickPochelu** [google/jax#13398]
Is possible to decompile the jaxpr code into a more readable Python code ? Python code is easier to edit than jaxpr code. Even if some information about the original function is lost (obfuscated code) like variable names being lost, it can be an important tool for reverse-engineering.

Here an illustration of the problem.

**(step a)**  f(x) is my function:
```
from jax import numpy as jnp
f=lambda x: jnp.log(1+jnp.exp(x)
```

**(step b)**  The derivative does not give the right answer:
`print(grad(f)(100.)) # nan <- expected 1`

**(step c)**  The JAXPR code of the derivative (along x axis) is:
```
from jax import make_jaxpr
make_jaxpr(grad(f))(100.)
```
output:
{ lambda ; a:f32[]. let
    b:f32[] = exp a
    c:f32[] = add 1.0 b
    _:f32[] = log c
    d:f32[] = div 1.0 c
    e:f32[] = mul d b
  in (e,) }

**(step d)** The Python equivalent (manually written) is:
```
df=lambda a: (1 / (1 + jnp.exp(a))) * jnp.exp(a)
print(df(100.)) # nan <- expected 1
```

**(step e)** I refactored a little bit the code and improved the arithmetic stability:
```
df=lambda x: 1 if x>10 else (jnp.exp(x) / (1 + jnp.exp(x)))
print(df(100.)) # 1. <- expected answer
```

A decompiler would automate step d.

**Comment by jakevdp** [comment]
Thanks for the suggestion! It's an interesting idea. There would be some complexity involved for a full solution; off the top of my head:

- how to decide if expressions are inlined or assigned to variables?
- how to deal with call primitives?
- how to deal with primitives whose API is quite different from the typical Python API, e.g. `scatter_p`, `gather_p`, `conv_general_dilated_p`, etc.?

It could be a fun challenge though 😁

**Comment by PierrickPochelu** [comment]
I will release a first version working for simple code samples in next days.

**Comment by jakevdp** [comment]
Cool - if you haven't seen it already, this might be a useful resource for crawling/interpreting jaxprs: https://jax.readthedocs.io/en/latest/notebooks/Writing_custom_interpreters_in_Jax.html

**Comment by PierrickPochelu** [comment]
I implemented a first version: https://github.com/PierrickPochelu/JaxDecompiler

For instance, it supports 23 common jaxpr operators such as "add, mul, neg, cos, sin,...". It supports also partially pmap.
